src/lib/react/macro/ReactTypeMacro.hx

```diff
@@ -106,93 +106,18 @@ class ReactTypeMacro
 		propsType:ComplexType,
 		stateType:ComplexType
 	) {
-		var partialStateType = switch (ComplexTypeTools.toType(stateType)) {
-			case TType(_):
-				TPath({
-					name: 'Partial',
-					pack: ['react'],
-					params: [TPType(stateType)]
-				});
-
-			default:
-				macro :Dynamic;
-		};
-
-		var setStateArgs:Array<FunctionArg> = [
-			{
-				name: 'nextState',
-				// TState -> Partial<TState>
-				type: TFunction([stateType], partialStateType),
-				opt: false
-			},
-			{
-				name: 'callback',
-				type: macro :Void->Void,
-				opt: true
-			}
-		];
-
-		fields.push({
-			name: 'setState',
-			access: [APublic, AOverride],
-			meta: [
-				{
-					// Add @:extern meta so that this code only exist at compile time
-					name: ':extern',
-					params: null,
-					pos: Context.currentPos()
-				},
-				{
-					// First overload:
-					// function(nextState:TState -> TProps -> Partial<TState>, ?callback:Void -> Void):Void {}
-					name: ':overload',
-					params: [generateSetStateOverload(
-						TFunction([stateType, propsType], partialStateType)
-					)],
-					pos: Context.currentPos()
-				},
-				{
-					// Second overload:
-					// function(nextState:Partial<TState>, ?callback:Void -> Void):Void {}
-					name: ':overload',
-					params: [generateSetStateOverload(partialStateType)],
-					pos: Context.currentPos()
-				}
-			],
-			kind: FFun({
-				args: setStateArgs,
-				ret: macro :Void,
-				#if haxe4
-				expr: null
+		fields.push((macro class C {
+			@:extern
+			@:overload(function(nextState:$stateType -> react.Partial<$stateType>, ?callback:Void -> Void):Void {})
+			@:overload(function(nextState:$stateType -> $propsType -> react.Partial<$stateType>, ?callback:Void -> Void):Void {})
+			override public function setState(state: react.Partial<$stateType>, ?callback:Void -> Void): Void
+				#if (haxe4 || haxe_ver > 4.0) // some preview builds required the function body and also didn't set the haxe4 flag
```

Comment:
`if haxe4` was intentional; the only preview build that did not want the function body was preview 5, which already had some issues with react ecosystem (mainly webpack). Many react projects were (and still are) using preview 4 and we shouldn't break for them.

If you want support for preview 5, there was support for `-D haxe4preview5`in `react-next` before rc1, you can add it back along `haxe4`.

Btw, you can inverse the logic in the if instead of using if/else with an empty if section.

Reply:
I think we're ok here – preview 4 doesn't set the `haxe4` flag (actually neither does preview 5) so a function body is generated. However, rc1 and the latest haxe nightly builds do define `haxe4` _and_ don't require a function body. 

So none of the previews are broken and will work with new releases of haxe (the `haxe_ver > 4.0` is so a function body isn't generated in the future if `haxe4` is dropped in say haxe 5).

The only weird one is preview5, which only warns about the function body but compiles ok, which I think is an acceptable situation for a preview build, but happy to add `haxe4preview5`.

I felt the positive condition was easier to understand than the inverted, especially when there's this headfake about `haxe4` not actually meaning all haxe 4 builds.

Reply:
> preview 4 doesn't set the `haxe4` flag

But preview 4 does set `haxe_ver` to `4.000`, that's why I don't want this check.

> in the future if `haxe4` is dropped in say haxe 5

`haxeN` flags are kept in future versions, haxe 5 will have `haxe3`, `haxe4` and `haxe5` flags.

Reply:
I was curious about that, yeah `haxe3` seems to be still kept, I'll remove the haxe_ver > 4.